### PR TITLE
ISSUE-212 LongLivedToken/set create

### PR DESCRIPTION
**Pull request by vttranlina**
Ref https://github.com/linagora/tmail-backend/issues/212

### CODE REVIEW

File: tmail-backend/jmap/extensions/src/test/scala/com/linagora/tmail/james/jmap/json/LongLivedTokenSerializerTest.scala [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+class LongLivedTokenSerializerTest {
+
+  @Test
+  def deserializeLongLivedTokenSetRequestShouldSuccess(): Unit = {
+    val jsInput: JsValue = Json.parse(
+      """{
+        |    "accountId": "aHR0cHM6Ly93d3cuYmFzZTY0ZW5jb2RlLm9yZy8",
+        |    "create": {
+        |        "deviceId": "My android device"
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/test/scala/com/linagora/tmail/james/jmap/json/LongLivedTokenSerializerTest.scala:24]: Please embed client generated id as per `/set create spec`. What happens when we want to create 2 tokens at once?

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/LongLivedTokenSetMethod.scala [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+class LongLivedTokenSetMethodMethodModule extends AbstractModule {
+  override def configure(): Unit = {
+    Multibinder.newSetBinder(binder(), classOf[Method])
+      .addBinding()
+      .to(classOf[LongLivedTokenSetMethod])
+
+    bind(classOf[InMemoryLongLivedTokenStore]).in(Scopes.SINGLETON)
+    bind(classOf[LongLivedTokenStore]).to(classOf[InMemoryLongLivedTokenStore])
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/LongLivedTokenSetMethod.scala:40]: Have the binding to the actual implementation done in another guice module please.

File: tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraLongLivedTokenSetMethodContract.scala [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+         |                    "token": "$${json-unit.ignore}"
+         |                }
+         |            },
+         |            "c1"
+         |        ]
+         |    ]
+         |}""".stripMargin)
+  }
+
+}
```
Comment (chibenwa) [inline tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraLongLivedTokenSetMethodContract.scala:374]: You are missing tests regarding:

 - creating two long lived tokens in a call
 - Referencing a long lived token using the creationId supplied by the client (prefixed by `#`) - use echo method
 - Using back references. - use echo method

File: tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraLongLivedTokenSetMethodContract.scala [hunk 4]
```diff
@@ -371,4 +524,75 @@ trait LinagoraLongLivedTokenSetMethodContract {
...
+         |                        "token": "$${json-unit.ignore}"
+         |                    }
+         |                }
+         |            },
+         |            "c1"
+         |        ],
+         |        [
+         |            "Core/echo",
+         |            {
+         |                "arg1": "$${json-unit.ignore}"
```
Comment (chibenwa) [inline tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraLongLivedTokenSetMethodContract.scala:578]: We could assert that arg1 value is not `#K38` ?

Comment (vttranlina) [reply tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraLongLivedTokenSetMethodContract.scala:578]: I think with this code (it already), it is ok
```scala
    val longLivedTokenId:String = (((Json.parse(response) \\ "methodResponses" )
      .head \\ "created")
      .head \\ "id")
      .head.asInstanceOf[JsString].value

    val backReferenceValue:String = ((Json.parse(response) \\ "methodResponses" )
      .head \\ "arg1")
      .head.asInstanceOf[JsString].value

    assertThat(longLivedTokenId)
      .isEqualTo(backReferenceValue)
```

Comment (chibenwa) [reply tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraLongLivedTokenSetMethodContract.scala:578]: Yes it is.

**Review by chibenwa — APPROVED** [review]

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/LongLivedTokenSetMethod.scala [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+case object LongLivedTokenCapabilityProperties extends CapabilityProperties {
+  override def jsonify(): JsObject = Json.obj()
+}
+
+case object LongLivedTokenCapability extends Capability {
+  val properties: CapabilityProperties = LongLivedTokenCapabilityProperties
+  val identifier: CapabilityIdentifier = LINAGORA_LONG_LIVED_TOKEN
+}
+
+class LongLivedTokenSetMethodMethodModule extends AbstractModule {
```
Comment (Arsnael) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/LongLivedTokenSetMethod.scala:34]:
```suggestion
class LongLivedTokenSetMethodModule extends AbstractModule {
```

File: tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraLongLivedTokenSetMethodContract.scala [hunk 6]
```diff
@@ -0,0 +1,598 @@
...
+         |            {
+         |                "type": "accountNotFound"
+         |            },
+         |            "c1"
+         |        ]]
+         |}""".stripMargin)
+  }
+
+  @Test
+  def methodShouldFailWhenDeviceIdTooLength(): Unit = {
```
Comment (Arsnael) [inline tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraLongLivedTokenSetMethodContract.scala:159]:
```suggestion
  def methodShouldFailWhenDeviceIdTooLong(): Unit = {
```

File: tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraLongLivedTokenSetMethodContract.scala [hunk 7]
```diff
@@ -0,0 +1,598 @@
...
+         |    "sessionState": "${SESSION_STATE.value}",
+         |    "methodResponses": [
+         |        [
+         |            "LongLivedToken/set",
+         |            {
+         |                "accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+         |                "notCreated": {
+         |                    "K38": {
+         |                        "type": "invalidArguments",
+         |                        "description": "deviceId must be not empty"
```
Comment (Arsnael) [inline tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraLongLivedTokenSetMethodContract.scala:247]:
```suggestion
         |                        "description": "deviceId must not be empty"
```

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/json/LongLivedTokenSerializer.scala [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+  private implicit val longLivedTokenSetRequestRead: Reads[LongLivedTokenSetRequest] = Json.reads[LongLivedTokenSetRequest]
+  private implicit val longLivedTokenSetResponseWrite: Writes[LongLivedTokenSetResponse] = Json.writes[LongLivedTokenSetResponse]
+
+  def deserializeLongLivedTokenSetRequest(input: JsValue): JsResult[LongLivedTokenSetRequest] =
+    Json.fromJson[LongLivedTokenSetRequest](input)
+
+  def deserializeLongLivedTokenCreationRequest(input: JsValue): JsResult[LongLivedTokenCreationRequest] =
+    Json.fromJson[LongLivedTokenCreationRequest](input)
+
+  def serializeKeystoreGetResponse(response: LongLivedTokenSetResponse): JsValue = Json.toJson(response)
```
Comment (Arsnael) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/json/LongLivedTokenSerializer.scala:37]: copy/paste? :)

`serializeKeystoreGetResponse` -> `serializeLongLivedTokenSetResponse`

**Review by Arsnael — APPROVED** [review]

**Pull request merged**